### PR TITLE
Support better inventory management, equipments, stats, items, synchronization, hit frames, elemental  &environmental damages, knockback

### PR DESCRIPTION
**Pull request by Sceat**
One size fit all PR for ease of development, it'll be divided into smaller other ones now that the main chunk of work is done.

## Core
- [x] typings
- [x] inventory
- [x] equipments
- [x] stats
- [x] secondary stats effects #640 (except `reach`)
- [x] items
- [x] sync
- [x] sneak
- [x] hit frames
- [x] elemental damages
- [x] naive knockback
- [x] environmental damages
- [x] fix mobs pathfinders

## UI

- [x] player head
- [x] health bar
- [x] health text
- [x] experience bar
- [x] experience text
- [x] level
- [ ] soul
- [ ] class
- [ ] group
- [ ] zone
- [ ] weapon
- [ ] spells
- [ ] consumable
- [ ] stats points remaining 

## GUI

- [x] items
- [ ] equipments
- [ ] stats
- [ ] settings

## External

- [ ] mobs name
- [ ] players titles

### CODE REVIEW

**Comment by Sceat** [comment]
Test failing while the ressourcepack repo is private, will be fixed

File: src/permissions.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { GameMode } from './gamemode.js'
+
+export function can_receive_damage({
+  game_mode = undefined,
+  soul = undefined,
+} = {}) {
```
Comment (DeltaEvo) [inline src/permissions.js:6]:
```suggestion
}) {
```

File: src/player/health.js [hunk 2]
```diff
@@ -46,18 +52,62 @@ export default {
       })
     })
 
+    aiter(abortable(setInterval(1000, null, { signal })))
+      .map(get_state)
+      .filter(state => !!state)
```
Comment (DeltaEvo) [inline src/player/health.js:57]: Nit: to be consistent with another part of the code

```suggestion
      ..dropWhile(state => !state)
```

File: src/player/login.js [hunk 3]
```diff
@@ -6,21 +8,66 @@ import { PLAYER_ENTITY_ID } from '../settings.js'
...
 import { set_world_border } from './world_border.js'
-import { BLOCK_TAGS, ENTITY_TAGS, FLUID_TAGS, ITEM_TAGS } from './tags.js'
+
+const log = logger(import.meta)
+const A_WEEK = 1000 * 60 * 60 * 24 * 7
+
+/** returns a 8x8 pixels matrix from an uuid */
+function fetch_head_pixels(client) {
+  log.info({ player: client.username }, `fetch head's pixels`)
+  return fetch(`https://minotar.net/avatar/${client.uuid}/8`)
```
Comment (DeltaEvo) [inline src/player/login.js:23]: Add a todo or create an issue, we should not depend on external services like that.

File: src/mobs/damage.js [hunk 4]
```diff
@@ -1,76 +1,202 @@
...
-          health,
-          last_hit: now,
-        }
+      return {
+        first_damager: damager,
+        ...state,
+        last_damager: damager,
+        last_hit_was_critical: critical_hit,
+        health,
+        path: [knockback_target_block],
```
[the remark this answers is not in the document]
Comment (Sceat) [reply src/mobs/damage.js:82]: it's a naive knockback but it's dogshit, better than nothing but there is a big chunk of work to do on mobs behaviors, it's messy for now they don't even have gravity

File: src/events.js [hunk 5]
```diff
@@ -4,66 +4,87 @@ export function last_event_value(emitter, event) {
...
-  DIE: 'PLAYER:ACTION:DIE',
-  /** the inventory of the player should be resynced */
-  RESYNC_INVENTORY: 'PLAYER:ACTION:RESYNC_INVENTORY',
-  /** the gamemode of the player should be updated */
-  SWITCH_GAMEMODE: 'PLAYER:ACTION:SWITCH_GAMEMODE',
-  /** the player should receive experience */
-  RECEIVE_EXPERIENCE: 'PLAYER:ACTION:RECEIVE_EXPERIENCE',
-}
-
+/** Local events which can be
```
[the remark this answers is not in the document]
Comment (Sceat) [reply src/events.js:7]: Idea was that we dispatch actions and emit events, but some need to be both so it's confusing. It's still confusing with the fact that sometimes we dispatch to reduce and some time we emit to observe, would be nice to merge those concept too somehow to simplify everything. 

we would just "dispatchemit" and handle wherever we need (observe/reduce)

File: src/player/damage.js [hunk 6]
```diff
@@ -88,21 +142,35 @@ export default {
...
+            damage > 0 ? Colors.DARK_RED : Colors.DARK_GREEN
 
           if (!is_dead) {
-            create_armor_stand(client, entity_id, damage_position, {
-              text: `-${damage}`,
-              color: '#E74C3C', // https://materialui.co/flatuicolors Alizarin
+            const sign = damage > 0 ? '-' : ''
+            create_armor_stand(client, entity_id, position, {
+              text: `${sign}${damage}`,
+              color: critical_hit ? critical_color : color,
```
[the remark this answers is not in the document]
Comment (Sceat) [reply src/player/damage.js:155]: `damage` is a value computed from raw damage, life steal, and heal, the result value will be negative if there is more heal on the current randomized hit

File: src/mobs/damage.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/mobs/damage.js:82]: Ack

File: src/events.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/events.js:7]: > but some need to be both so it's confusing.

This should never happens for me as it's two different things, things that are dispatched trigger a mutation where events are merely here to decouple the code (We could in theory replace all emits by function calls)

I think we need this distinction, as events should always stay local.

File: src/player/damage.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/player/damage.js:155]: Ack

[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/player/damage.js:155]: Wait, in this case you have a bug, don't you want `text: ${sign}{Math.abs(damage)}` ?

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**